### PR TITLE
8302113: Improve CRC32 intrinsic with crypto pmull on AArch64

### PR DESCRIPTION
**Pull request by yftsai**
Instruction pmull and pmull2 support operating on 64-bit data in Cryptographic Extension. The execution throughput of this form raises from 1 on Neoverse N1 to 4 on Neoverse V1 while the latency remains 2. The CRC32 instructions did not changed: latency 2, throughput 1. As a result, computing CRC32 using pmull could perform better than using crc32 instruction.

The following test has passed.
test/hotspot/jtreg/compiler/intrinsics/zip/TestCRC32.java

The throughput reported by [the micro benchmark](https://github.com/openjdk/jdk/blob/master/test/micro/org/openjdk/bench/java/util/TestCRC32.java) is measured on an EC2 c7g instance. The optimization shows 11 - 99% improvement when the input is at least 384 bytes.

| input               | 64         | 128        | 256        | 384        | 511        | 512        | 1,024      |
| ------------- | ---------- | ---------- | ---------- | ---------- | ---------- | ---------- | ---------- |
|  improvement  | 0.02%      | 0.02%      | 0.00%      | 16.00%     | 11.94%     | 34.75%     | 69.80%     |

| input               | 2,048      | 4,096      | 8,192      | 16,384     | 32,768     | 65,536     |
| ------------- | ---------- | ---------- | ---------- | ---------- | ---------- | ---------- |
|  improvement  | 77.61%     | 92.33%     | 95.98%     | 97.95%     | 99.33%     | 98.36%     |

Baseline
```
TestCRC32.testCRC32Update         64  thrpt   12  173126.358 ± 118.330  ops/ms
TestCRC32.testCRC32Update        128  thrpt   12  112910.118 ±  47.305  ops/ms
TestCRC32.testCRC32Update        256  thrpt   12   66601.990 ±   7.294  ops/ms
TestCRC32.testCRC32Update        384  thrpt   12   47229.319 ±   3.949  ops/ms
TestCRC32.testCRC32Update        511  thrpt   12   33733.119 ±   4.076  ops/ms
TestCRC32.testCRC32Update        512  thrpt   12   36584.565 ±   4.211  ops/ms
TestCRC32.testCRC32Update       1024  thrpt   12   19239.083 ±   1.040  ops/ms
TestCRC32.testCRC32Update       2048  thrpt   12    9875.652 ±   0.435  ops/ms
TestCRC32.testCRC32Update       4096  thrpt   12    5004.425 ±   0.290  ops/ms
TestCRC32.testCRC32Update       8192  thrpt   12    2519.185 ±   0.169  ops/ms
TestCRC32.testCRC32Update      16384  thrpt   12    1263.909 ±   0.194  ops/ms
TestCRC32.testCRC32Update      32768  thrpt   12     632.018 ±   0.053  ops/ms
TestCRC32.testCRC32Update      65536  thrpt   12     315.471 ±   0.095  ops/ms
```

Crypto pmull
```
TestCRC32.testCRC32Update         64  thrpt   12  173168.669 ±   4.746  ops/ms
TestCRC32.testCRC32Update        128  thrpt   12  112933.519 ±   4.583  ops/ms
TestCRC32.testCRC32Update        256  thrpt   12   66602.462 ±   3.150  ops/ms
TestCRC32.testCRC32Update        384  thrpt   12   54784.739 ±   2.110  ops/ms
TestCRC32.testCRC32Update        511  thrpt   12   37760.816 ±  69.911  ops/ms
TestCRC32.testCRC32Update        512  thrpt   12   49297.609 ±  21.983  ops/ms
TestCRC32.testCRC32Update       1024  thrpt   12   32667.507 ±  90.610  ops/ms
TestCRC32.testCRC32Update       2048  thrpt   12   17539.986 ± 511.416  ops/ms
TestCRC32.testCRC32Update       4096  thrpt   12    9625.249 ±   9.713  ops/ms
TestCRC32.testCRC32Update       8192  thrpt   12    4937.135 ±   6.121  ops/ms
TestCRC32.testCRC32Update      16384  thrpt   12    2501.936 ±   1.270  ops/ms
TestCRC32.testCRC32Update      32768  thrpt   12    1259.831 ±   0.119  ops/ms
TestCRC32.testCRC32Update      65536  thrpt   12     625.773 ±   0.242  ops/ms
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8302113](https://bugs.openjdk.org/browse/JDK-8302113): Improve CRC32 intrinsic with crypto pmull on AArch64

### Reviewers
 * [Volker Simonis](https://openjdk.org/census#simonis) (@simonis - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12480/head:pull/12480` \
`$ git checkout pull/12480`

Update a local copy of the PR: \
`$ git checkout pull/12480` \
`$ git pull https://git.openjdk.org/jdk pull/12480/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12480`

View PR using the GUI difftool: \
`$ git pr show -t 12480`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12480.diff">https://git.openjdk.org/jdk/pull/12480.diff</a>

</details>

### CODE REVIEW

**Comment by eastig** [comment]
In the description, could you please combine data for the baseline and the optimized versions into a table? It's hard to compare them as they are now.

**Comment by eastig** [comment]
How was the correctness of the implementation tested?
What a microbenchmark did you use? It is not mentioned in the description.
There is `test/micro/org/openjdk/bench/java/util/TestCRC32C.java`. Should we add a similar benchmark for CRC32?

**Comment by yftsai** [comment]
> How was the correctness of the implementation tested? What a microbenchmark did you use? It is not mentioned in the description. There is `test/micro/org/openjdk/bench/java/util/TestCRC32C.java`. Should we add a similar benchmark for CRC32?

The micro benchmark has been added recently. The performance measurement is slightly different, so the threshold is raised to 384 bytes.

**Comment by eastig** [comment]
> [Support CRC32-C](https://github.com/openjdk/jdk/pull/12480/commits/1cc1e55faf3bb5fcffb5fb7cc8851f25ea208aa4)

Could you please create a separate PR for CRC32C?

File: src/hotspot/cpu/aarch64/globals_aarch64.hpp [hunk 1]
```diff
@@ -88,6 +88,8 @@ define_pd_global(intx, InlineSmallCode,          1000);
           "Use Neon for CRC32 computation")                             \
   product(bool, UseCRC32, false,                                        \
           "Use CRC32 instructions for CRC32 computation")               \
+  product(bool, UseCryptoPmull, false,                                  \
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/globals_aarch64.hpp:91]:
```suggestion
  product(bool, UsePmullForCRC32, false,                                  \
```

**Comment by theRealAph** [comment]
Why do we need any new code for this processor? We already have routines for CRC32 using pmull.

**Comment by yftsai** [comment]
> Why do we need any new code for this processor? We already have routines for CRC32 using pmull.

Do you refer to [this pmull implementation](https://github.com/openjdk/jdk/blob/1ef9f6507ba45419f0fa896915eec064762c5153/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L3753)? It uses pmull to take in 8-bit elements and produces 16-bit results.
The new code uses pmull variation in Cryptographic Extension to take in 32-bit elements and produces 64-bit results. It is more efficient along with eor3.

**Comment by theRealAph** [comment]
> > Why do we need any new code for this processor? We already have routines for CRC32 using pmull.
> 
> Do you refer to [this pmull implementation](https://github.com/openjdk/jdk/blob/1ef9f6507ba45419f0fa896915eec064762c5153/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L3753)? It uses pmull to take in 8-bit elements and produces 16-bit results. The new code uses pmull variation in Cryptographic Extension to take in 32-bit elements and produces 64-bit results. It is more efficient along with eor3.

How much more efficient is it than the old version of CRC32 using pmull? And do we today need three different hand-coded versions of CRC32?

**Comment by yftsai** [comment]
> How much more efficient is it than the old version of CRC32 using pmull?
The old version of CRC32 using pmull is inefficient comparing to the version using crc32 instructions measured on the same EC2 c7g instance.

The following is the performance comparing to the version using crc32 instructions.
| input                              | 64         | 128        | 256        | 384        | 511        | 512        | 1,024      |
| ---------------------------------- | ---------- | ---------- | ---------- | ---------- | ---------- | ---------- | ---------- |
| -XX:-UseCRC32 -XX:+UseNeon (pmull) | -88.10%    | -88.31%    | -88.57%    | -88.64%    | -88.71%    | -88.73%    | -88.85%    |
| -XX:-UseCRC32 -XX:-UseNeon         | -90.45%    | -92.53%    | -93.62%    | -93.98%    | -93.73%    | -94.17%    | -94.51%    |

| input                              | 2,048      | 4,096      | 8,192      | 16,384     | 32,768     | 65,536     |
| -----------------------------------| ---------- | ---------- | ---------- | ---------- | ---------- | ---------- |
| -XX:-UseCRC32 -XX:+UseNeon (pmull) | -88.92%    | -88.96%    | -88.97%    | -88.98%    | -88.96%    | -88.97%    |
| -XX:-UseCRC32 -XX:-UseNeon         | -94.58%    | -94.65%    | -94.68%    | -94.70%    | -94.70%    | -94.70%    |
  

`-XX:-UseCRC32 -XX:+UseNeon` (pmull)
```
TestCRC32.testCRC32Update         64  thrpt   12   20593.817 ±   8.676  ops/ms
TestCRC32.testCRC32Update        128  thrpt   12   13197.522 ±   4.268  ops/ms
TestCRC32.testCRC32Update        256  thrpt   12    7610.119 ±   6.329  ops/ms
TestCRC32.testCRC32Update        384  thrpt   12    5365.177 ±   0.570  ops/ms
TestCRC32.testCRC32Update        511  thrpt   12    3808.462 ±   3.441  ops/ms
TestCRC32.testCRC32Update        512  thrpt   12    4121.495 ±   4.549  ops/ms
TestCRC32.testCRC32Update       1024  thrpt   12    2144.889 ±   1.610  ops/ms
TestCRC32.testCRC32Update       2048  thrpt   12    1093.756 ±   0.114  ops/ms
TestCRC32.testCRC32Update       4096  thrpt   12     552.393 ±   0.164  ops/ms
TestCRC32.testCRC32Update       8192  thrpt   12     277.963 ±   0.099  ops/ms
TestCRC32.testCRC32Update      16384  thrpt   12     139.233 ±   0.061  ops/ms
TestCRC32.testCRC32Update      32768  thrpt   12      69.752 ±   0.013  ops/ms
TestCRC32.testCRC32Update      65536  thrpt   12      34.789 ±   0.002  ops/ms
```

`-XX:-UseCRC32 -XX:-UseNeon`
```
TestCRC32.testCRC32Update         64  thrpt   12   16541.230 ±   50.847  ops/ms
TestCRC32.testCRC32Update        128  thrpt   12    8432.515 ±    6.676  ops/ms
TestCRC32.testCRC32Update        256  thrpt   12    4252.450 ±    2.897  ops/ms
TestCRC32.testCRC32Update        384  thrpt   12    2844.691 ±    0.110  ops/ms
TestCRC32.testCRC32Update        511  thrpt   12    2114.317 ±    2.670  ops/ms
TestCRC32.testCRC32Update        512  thrpt   12    2134.668 ±    0.150  ops/ms
TestCRC32.testCRC32Update       1024  thrpt   12    1055.619 ±   27.552  ops/ms
TestCRC32.testCRC32Update       2048  thrpt   12     535.514 ±    0.273  ops/ms
TestCRC32.testCRC32Update       4096  thrpt   12     267.859 ±    0.133  ops/ms
TestCRC32.testCRC32Update       8192  thrpt   12     133.916 ±    0.014  ops/ms
TestCRC32.testCRC32Update      16384  thrpt   12      67.003 ±    0.034  ops/ms
TestCRC32.testCRC32Update      32768  thrpt   12      33.471 ±    0.023  ops/ms
TestCRC32.testCRC32Update      65536  thrpt   12      16.724 ±    0.018  ops/ms

> And do we today need three different hand-coded versions of CRC32?
Each version depends on different processor features. Removing the least efficient versions could degrade existing use cases.
It might be still reasonable to remove them though. The crc32c intrinsic only has one version, which uses crc32c instruction.

**Comment by theRealAph** [comment]
The optional CRC instructions in v8.0 become a requirement in ARMv8.1. ARMv8.0 is stuff like Cortex A53, but also including, apparently, Cortex A72. So I guess we're stuck with all three, at least for now. This is a bad business, but I guess it's something that Arm & partners have dropped on us, and there's little we can do about it.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 2]
```diff
@@ -3715,7 +3774,11 @@ void MacroAssembler::kernel_crc32(Register crc, Register buf, Register len,
   Label L_by16, L_by16_loop, L_by4, L_by4_loop, L_by1, L_by1_loop, L_exit;
 
   if (UseCRC32) {
-      kernel_crc32_using_crc32(crc, buf, len, table0, table1, table2, table3);
+      if (UseCryptoPmullForCRC32) {
```
Comment (simonis) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:3777]: Whis is this inside `UseCRC32`? It means that if we set `-XX:-UseCRC32 -XX:+UseCryptoPmullForCRC32` on the command line, we won't get `UseCryptoPmullForCRC32`

File: src/hotspot/cpu/aarch64/vm_version_aarch64.cpp [hunk 3]
```diff
@@ -246,6 +246,19 @@ void VM_Version::initialize() {
...
 
+
+  // Neoverse V1
+  if (_cpu == CPU_ARM && (_model == 0xd40 || _model2 == 0xd40)) {
+    if (FLAG_IS_DEFAULT(UseCryptoPmullForCRC32)) {
+      FLAG_SET_DEFAULT(UseCryptoPmullForCRC32, true);
+    }
+  }
+
+  if (UseCryptoPmullForCRC32 && (!VM_Version::supports_sha3() || !VM_Version::supports_pmull())) {
```
Comment (simonis) [inline src/hotspot/cpu/aarch64/vm_version_aarch64.cpp:257]: Why does `UseCryptoPmullForCRC32` depend on `VM_Version::supports_sha3()`?

Does `VM_Version::supports_pmull()` really check for the `pmull` in the cryptographic extension?

**Review by simonis — COMMENTED** [review]
In general your changes look good. I only have a few minor questions/comments below.

Also, can you please update the copyright year in the files you've touched?

Thank you and best regards,
Volker

File: src/hotspot/cpu/aarch64/globals_aarch64.hpp (hunk 1 above)
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/globals_aarch64.hpp]: > Changed to UseCryptoPmullForCRC32 as there is another pmull version.

Can you give me a reference for that? I can only find one definition of pmull.

File: src/hotspot/cpu/aarch64/vm_version_aarch64.cpp (hunk 3 above)
Comment (yftsai) [reply src/hotspot/cpu/aarch64/vm_version_aarch64.cpp:257]: > Why does `UseCryptoPmullForCRC32` depend on `VM_Version::supports_sha3()`?

The implementation uses [eor3](https://developer.arm.com/documentation/ddi0602/2022-06/SIMD-FP-Instructions/EOR3--Three-way-Exclusive-OR-) as well, and it is only available when SHA3 is implemented.
 
> Does `VM_Version::supports_pmull()` really check for the `pmull` in the cryptographic extension?

supports_pmull() returns true when CPU_PMULL/HWCAP_PMULL of VM_Version::_features is [enabled](https://github.com/openjdk/jdk/blob/3ba156082b73c4a8e9d890a57a42fb68df2bf98f/src/hotspot/cpu/aarch64/vm_version_aarch64.hpp#L121). HWCAP_PMULL refers to the [functionality](https://docs.kernel.org/arm64/elf_hwcaps.html) implied by ID_AA64ISAR0_EL1.AES == 0b0010, i.e. [PMULL/PMULL2](https://developer.arm.com/documentation/ddi0595/2021-03/AArch64-Registers/ID-AA64ISAR0-EL1--AArch64-Instruction-Set-Attribute-Register-0) instructions operating on 64-bit data quantities.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp (hunk 2 above)
Comment (yftsai) [reply src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp]: crc32 instructions are used to process remaining input not multiple of 128 bytes. The dependency between these two options is removed.

File: src/hotspot/cpu/aarch64/globals_aarch64.hpp (hunk 1 above)
Comment (yftsai) [reply src/hotspot/cpu/aarch64/globals_aarch64.hpp]: There will be two pmull implementations
1) `-XX:+UseCryptoPmullForCRC32` to use this 64x64 pmull implementation
2) `-XX:-UseCryptoPmullForCRC32 -XX:-UseCRC32 -XX:+UseNeon` to use the [8x8 pmull implementation](https://github.com/openjdk/jdk/blob/1ef9f6507ba45419f0fa896915eec064762c5153/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L3753).

**Review by simonis — APPROVED** [review]
Looks good now. Thanks for the explanation and for decoupling `UseCryptoPmullForCRC32` and `UseCRC32`.

**Pull request closed without merging**